apps/webapp/app/v3/marqs.server.ts

```diff
@@ -0,0 +1,906 @@
+  }
+
+  // This should increment by the number of seconds, but with a max value of Date.now() + visibilityTimeoutInMs
+  public async heartbeatMessage(messageId: string, seconds: number = 30) {
+    await this.#callHeartbeatMessage({
+      visibilityQueue: constants.MESSAGE_VISIBILITY_TIMEOUT_QUEUE,
+      messageId,
+      milliseconds: seconds * 1000,
+      maxVisibilityTimeout: Date.now() + this.visibilityTimeoutInMs,
+    });
+  }
+
+  get visibilityTimeoutInMs() {
+    return this.options.visibilityTimeoutInMs ?? 300000;
+  }
+
+  async #readMessage(messageId: string) {
+    return this.#trace(
+      "readMessage",
+      async (span) => {
+        const rawMessage = await this.redis.get(`${constants.MESSAGE_PART}:${messageId}`);
+
+        if (!rawMessage) {
+          return;
+        }
+
+        const message = MessagePayload.safeParse(JSON.parse(rawMessage));
+
+        if (!message.success) {
+          return;
+        }
+
+        return message.data;
+      },
+      { attributes: { [SemanticAttributes.MESSAGE_ID]: messageId } }
+    );
+  }
+
+  async #getRandomQueueFromParentQueue(
+    parentQueue: string,
+    calculateWeight: (queue: string, score: number) => Promise<number>
+  ) {
+    return this.#trace(
+      "getRandomQueueFromParentQueue",
+      async (span, abort) => {
+        const queues = await this.#zrangeWithScores(parentQueue, 0, -1);
+
+        if (queues.length === 0) {
+          abort();
+          return;
+        }
+
+        span.setAttribute("marqs.queueCount", queues.length);
+
+        const queuesWithWeights = await this.#calculateQueueWeights(queues, calculateWeight);
+
+        // We need to priority shuffle here to ensure all workers aren't just working on the highest priority queue
+        return await this.#weightedRandomChoice(queuesWithWeights);
+      },
+      { kind: SpanKind.CONSUMER, attributes: { [SemanticAttributes.PARENT_QUEUE]: parentQueue } }
+    );
+  }
+
+  // Calculate the weights of the queues based on the age and the capacity
+  async #calculateQueueWeights(
+    queues: Array<{ value: string; score: number }>,
+    calculateWeight: (queue: string, score: number) => Promise<number>
+  ) {
+    const queueWeights = await Promise.all(
+      queues.map(async (queue) => {
+        return {
+          queue: queue.value,
+          weight: await calculateWeight(queue.value, queue.score),
+        };
+      })
+    );
+
+    return queueWeights;
+  }
+
+  async #calculateMessageQueueWeight(queue: string, score: number) {
+    const concurrencyQueueName = queue.replace(/:ck:.+$/, "");
+
+    const concurrencyLimit =
+      (await this.redis.get(`${concurrencyQueueName}:${constants.CONCURRENCY_LIMIT_PART}`)) ?? 100;
+    const currentConcurrency = await this.redis.scard(
+      `${queue}:${constants.CURRENT_CONCURRENCY_PART}`
+    );
+
+    const capacity = Number(concurrencyLimit) - Number(currentConcurrency);
+
+    const capacityWeight = capacity / Number(concurrencyLimit);
```

Comment:
We should guard against the concurrencyLimit being zero (or negative).

If it's zero then `capacityWeight` will be Infinity.